plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java

```diff
@@ -623,4 +640,99 @@ private static void validateOrcBloomFilterColumns(ConnectorTableMetadata tableMe
+        String metadataDirectoryLocation = format("%s/%s", stripTrailingSlash(location), METADATA_FOLDER_NAME);
+        try {
+            int latestMetadataVersion = -1;
+            FileIterator fileIterator = fileSystem.listFiles(metadataDirectoryLocation);
+            while (fileIterator.hasNext()) {
+                FileEntry fileEntry = fileIterator.next();
+                if (fileEntry.path().endsWith(METADATA_FILE_EXTENSION)) {
+                    OptionalInt version = parseVersion(fileEntry.path());
+                    if (version.isPresent()) {
+                        if (version.getAsInt() > latestMetadataVersion) {
+                            latestMetadataVersion = version.getAsInt();
+                            latestMetadataLocations.clear();
+                            latestMetadataLocations.add(fileEntry.path());
+                        }
+                        else if (version.getAsInt() == latestMetadataVersion) {
+                            latestMetadataLocations.add(fileEntry.path());
+                        }
+                    }
+                }
+            }
+            if (latestMetadataLocations.isEmpty()) {
+                throw new TrinoException(ICEBERG_INVALID_METADATA, "No metadata file exists at location: " + metadataDirectoryLocation);
+            }
+            if (latestMetadataLocations.size() > 1) {
+                throw new TrinoException(ICEBERG_INVALID_METADATA, format("More than one latest metadata file found at location: %s, latest metadata files are %s",
+                        metadataDirectoryLocation, latestMetadataLocations));
+            }
+        }
+        catch (IOException e) {
+            throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, "Failed checking table's location: " + location, e);
+        }
+        return Optional.of(getOnlyElement(latestMetadataLocations.iterator()));
+    }
+
+    /**
+     * Get the latest metadata file location present in location if metadataFileName is not provided, otherwise
+     * form the metadata file location using location and metadataFileName
+     *
+     * @param fileSystem
+     * @param location
+     * @param metadataFileName
+     * @return full qualified metadata file location
+     */
+    public static Optional<String> getMetadataLocation(TrinoFileSystem fileSystem, String location, Optional<String> metadataFileName)
+    {
+        if (metadataFileName.isEmpty()) {
+            return getLatestMetadataLocation(fileSystem, location);
+        }
+        else {
+            return Optional.of(format("%s/%s/%s", stripTrailingSlash(location), METADATA_FOLDER_NAME, metadataFileName.get()));
+        }
+    }
+
+    private static OptionalInt parseVersion(String metadataLocation)
+    {
+        int versionStart = metadataLocation.lastIndexOf('/') + 1; // if '/' isn't found, this will be 0
+        int versionEnd = metadataLocation.indexOf('-', versionStart);
```

Comment:
what if `/` isn't found?

Reply:
Here is the observation when `/` or `-` or both isn't found

		parseVersion("0001-hhee.metadata.json"); // Optinal.of(1)
		parseVersion("/var/lib/0001hhee.metadata.json"); // OptionalInt.empty()
		parseVersion("0001hhee.metadata.json"); // OptionalInt.empty()

Reply:
to me it wasn't obvious that it works; if this stays, i'd advocate for having unit tests for this method

Reply:
This is where `IndexOutOfBoundsException` comes into the picture. When `/` or `-` is not found and versionStart and versionEnd will not have the correct value then it is throwing `IndexOutOfBoundsException`

Reply:
Added unit tests `TestIcebergUtil`